clouddriver-google-common/src/main/groovy/com/netflix/spinnaker/clouddriver/googlecommon/deploy/GoogleCommonSafeRetry.groovy

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.googlecommon.deploy
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import groovy.util.logging.Slf4j
+
+import java.util.concurrent.TimeUnit
+
+@Slf4j
+class GoogleCommonSafeRetry {
+  /**
+   * Retry an operation if it fails. Treat any error codes in successfulErrorCodes as success.
+   *
+   * @param operation - The operation.
+   * @param action - String describing the operation.
+   * @param resource - Resource we are operating on.
+   * @param task - Spinnaker task. Can be null.
+   * @param phase
+   * @param retryCodes - GoogleJsonResponseException codes we retry on.
+   * @param successfulErrorCodes - GoogleJsonException codes we treat as success.
+   *
+   * @return Object returned from the operation.
+   */
+  public Object doRetry(Closure operation,
+                        String action,
+                        String resource,
+                        Task task,
+                        String phase,
+                        List<Integer> retryCodes,
+                        List<Integer> successfulErrorCodes,
+                        Class ProviderOperationException,
```

Comment:
The logic around throwing this exception type might read easier if you drop the reflection in favor of having some abstract exception factory method defined on the base class.

You wouldn't need to pass the `Class` arg and you could just call `throw providerOperationException('my message...')` or something similar.